akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -778,9 +788,16 @@ import scala.util.control.NonFatal
     status.compareAndSet(
       null,
       ActorSubscriberMessage.OnError(
-        new SubscriptionTimeoutException(s"Substream Source has not been materialized in $d")))
+        new SubscriptionTimeoutException(s"Substream Source($name) has not been materialized in $d")))
 
   override def createLogic(inheritedAttributes: Attributes) = new GraphStageLogic(shape) with OutHandler {
+    // check for previous materialization eagerly so we fail with a more useful stacktrace
+    private[this] val materializationException: OptionVal[IllegalStateException] =
+      if (status.get.isInstanceOf[AsyncCallback[_]])
+        OptionVal.Some(createMaterializedTwiceException())
+      else
+        OptionVal.None
```

Comment:
Couldn't it just throw here right away though?

Reply:
It did before, but then I changed it again. See https://github.com/akka/akka/pull/28492#issuecomment-574635642

Reply:
Ah, avoiding dangling sounds like it makes sense to me.